1-Draft/RFCNNNN-#Requires-Additions.md

```diff
@@ -0,0 +1,155 @@
+---
+RFC: RFC
+Author: Robert Holt
+Status: Draft
+SupercededBy:
+Version:
+Area: Hash-requires
+Comments Due: 2018-05-15
+Plan to implement: Yes
+---
+
+# \#Requires Additions
+
+Currently, PowerShell's `#requires` statement (or perhaps
+pragma?) supports the following parameters:
+
+* `-Version <N>[.<n>]`, where a minimum PowerShell version can be specified
+* `-PSSnapin <PSSnapin-Name> [-Version <N><n>]`, where a required PowerShell Snapin can be specified
+* `-Modules { <Module-name> | <Hashtable> }`, where PowerShell modules that are required can be specified
+* `-ShellId <ShellId>`, where the required Shell ID can be specified
+* `-RunAsAdministrator`, where the script is required to be run as administrator
+
+These features are documented in [about_Requires](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_requires?view=powershell-6), along with the statement that:
+
+> You can use #Requires statements in any script. You
+> cannot use them in functions, cmdlets, or snap-ins.
+
+Currently however, this is untrue, as `#requires`
+statements are allowed by the parser/tokenizer anywhere in
+a script and then effectively hoisted to the top of that
+script to be checked before any part of the script is
+executed, no matter where they are placed.
+
+This RFC proposes the following changes:
+
+* Only allow `#requires` at the top level of a script,
+  before any lines that are not comments (i.e. with the
+  intention that a hashbang can still work, just before
+  any executable PowerShell code). Placing `#requires` anywhere
+  after will cause a parse-time error. This would be a **breaking
+  change**, albeit one that the documentation already claims to be 
+  in force.
+* Using `#requires` in the interactive console will cause
+  a parse-time error. This could be a **minor breaking
+  change**, since currently PowerShell throws a [pipeline
+  creation error](https://github.com/PowerShell/PowerShell/issues/3803).
+* Add support for the following new parameters (each
+  independently up for discussion):
+  * `-OS {Windows | Linux | MacOS}`, where an
+    operating system (or possibly combination of them) can
+    be specified as required. See [this PowerShell issue](https://github.com/PowerShell/PowerShell/issues/3751).
+  * `-Assembly <Assembly-name>`, where a .NET assembly can
+    be specified as required. See [this PowerShell issue](https://github.com/PowerShell/PowerShell/issues/5022).
+  * `-MaxVersion <V>[.<v>]`, where a maximum PowerShell
```

Comment:
I don't know if we need a `-MaxVersion`.  The original issue was trying to differentiate Windows PowerShell and PowerShell Core.  Since these changes wouldn't go down to Windows PowerShell, it seems to be of limited use to use `-MaxVersion` to differentiate between the two.

Reply:
I think it does provide some benefit. Even though a version of PSCore is only supported for 18 months? scripts will be written in that time for that version. If we implement breaking changes in 6.3.0, a script written during 6.1.0's lifetime could put a max version of 6.1.0 in it to ensure that anyone running an older version of the script on a newer 6.3.0 version of pwsh, they will fail fast rather than risk potential broken behavior.

This goes to the larger discussion of how to deal with breaking changes, but I think this feature actually makes it possible to do them without major version madness.

Reply:
@markekraus for your scenario, it seems that an -ExactVersion would be more useful if you have business critical scripts that you want time to verify before moving to a newer version of PSCore6

Reply:
`-ExactVersion` would. for some instances. I think if we are committed to X.Y.Z versions where Z changes never include a breaking change, something that somehow limiting to a minor (Y) version would work and probably be what is desirable for most cases.  Basically way to do `(version >= 6.X.0 && version > 6.(X+1).0)`.

Reply:
If we say about business critical scripts I'd want to see error events in logs if the directive catch wrong version.

Reply:
If we'll want the `-MaxVersion` and `-ExactVersion`  I'd re-use `-Version` with prefixes like `<`,`>`, ...  - `-Version <=6.1.0`

Reply:
@iSazonov That would definitely make it more flexible... but... something like `-ExactVersion 6.0.2` and `-MinorVersion 6.1`   would cover probably most use cases (assuming we don't suddenly start making breaking changes in the PATCH level iterations)